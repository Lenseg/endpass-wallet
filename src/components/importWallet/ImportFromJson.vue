<template>
  <v-form>
    <div class="field">
      <div class="file">
        <label class="file-label">
          <input class="file-input"
                 type="file"
                 name="jsonWallet"
                 @change="setFile">
          <span class="file-cta">
            <span class="file-icon">
              <span class="icon is-small"
                    v-html="require('@/img/arrow-thick-top.svg')"></span>
            </span>
            <span class="file-label">
              {{ fileName || 'V3 JSON keystore file'}}
            </span>
          </span>
        </label>
      </div>
      <p v-show="errors.has('fileName')"
         class="help is-danger">{{ errors.first('fileName') }}</p>
    </div>

    <v-input v-model="jsonKeystorePassword"
             label="V3 JSON keystore password"
             id="jsonKeystorePassword"
             name="jsonKeystorePassword"
             type="password"
             validator="required|min:8"
             data-vv-as="password"
             key="jsonKeystorePasswordUnique"
             aria-describedby="jsonKeystorePassword"
             placeholder="V3 JSON keystore password"
             required />

    <v-button className="is-primary is-cta"
              :loading="isCreating"
              @click.prevent="parseJson">Import</v-button>
  </v-form>
</template>

<script>
import EthWallet from 'ethereumjs-wallet';
import router from '@/router';
import { mapActions, mapMutations } from 'vuex';
import VForm from '@/components/ui/form/VForm.vue';
import VInput from '@/components/ui/form/VInput.vue';
import VButton from '@/components/ui/form/VButton.vue';

export default {
  name: 'import-from-json',
  data: () => ({
    isCreating: false,
    jsonKeystorePassword: '',
    fileName: '',
    file: null,
  }),
  methods: {
    ...mapActions('accounts', ['addWalletWithV3']),
    parseJson() {
      const reader = new FileReader();
      reader.onload = this.addWalletWithJson.bind(this);

      try {
        reader.readAsText(this.file);
      } catch (e) {
        this.errors.add({
          field: 'fileName',
          msg: 'File is invalid',
          id: 'wrongFile',
        });
      }
    },
    async addWalletWithJson(e) {
      this.isCreating = true;

      await new Promise(res => setTimeout(res, 20));

      try {
        this.addWalletWithV3({
          json: e.target.result,
          password: this.jsonKeystorePassword,
        });
        router.push('/');
      } catch (e) {
        let error = {
          field: 'jsonKeystorePassword',
          msg: 'JSON password is invalid',
          id: 'wrongPass',
        };

        if (
          e.message === 'Not a V3 wallet' ||
          e.message.includes('Unexpected token')
        ) {
          error = {
            field: 'fileName',
            msg: 'File is invalid',
            id: 'wrongFile',
          };
        }

        this.errors.add(error);
      }

      this.isCreating = false;
    },
    setFile(e) {
      this.errors.removeById('wrongFile');

      if (e.target.files[0]) {
        this.fileName = e.target.files[0].name;
        this.file = e.target.files[0];
      } else {
        this.fileName = '';
        this.file = null;
      }
    },
  },
  watch: {
    jsonKeystorePassword() {
      this.errors.removeById('wrongPass');
    },
  },
  components: {
    VForm,
    VInput,
    VButton,
  },
};
</script>

<style lang="scss">
</style>
