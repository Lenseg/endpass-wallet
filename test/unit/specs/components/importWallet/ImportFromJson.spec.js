import { shallow, createLocalVue } from '@vue/test-utils';
import VeeValidate from 'vee-validate';

const localVue = createLocalVue();

localVue.use(VeeValidate);

import ImportFromJson from '@/components/importWallet/ImportFromJson';

describe('ImportFromJson', () => {
  describe('render', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(ImportFromJson, {
        localVue,
      });
    });

    it('should be a Vue component', () => {
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
