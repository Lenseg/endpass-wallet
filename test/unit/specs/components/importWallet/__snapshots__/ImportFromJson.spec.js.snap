// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportFromJson render should match snapshot 1`] = `<!---->`;
